ostd/src/mm/dma/dma_coherent.rs

```diff
@@ -1,186 +1,177 @@
-    /// or not.
+    /// Allocates a region of physical memory for coherent DMA access.
     ///
-    /// The method fails if any part of the given `segment`
-    /// already belongs to a DMA mapping.
-    pub fn map(segment: USegment, is_cache_coherent: bool) -> core::result::Result<Self, DmaError> {
-        let paddr = segment.paddr();
-        let frame_count = segment.size() / PAGE_SIZE;
-
-        if !check_and_insert_dma_mapping(paddr, frame_count) {
-            return Err(DmaError::AlreadyMapped);
-        }
+    /// If the device can access the memory with coherent access to the CPU
+    /// cache, set `is_cache_coherent` to `true`.
+    pub fn alloc(nframes: usize, is_cache_coherent: bool) -> Result<Self, Error> {
+        let has_tdx = has_tdx();
 
-        if !is_cache_coherent {
-            let page_table = KERNEL_PAGE_TABLE.get().unwrap();
-            let vaddr = paddr_to_vaddr(paddr);
-            let va_range = vaddr..vaddr + (frame_count * PAGE_SIZE);
-            // SAFETY: the physical mappings is only used by DMA so protecting it is safe.
-            unsafe {
-                page_table
-                    .protect_flush_tlb(&va_range, |p| p.cache = CachePolicy::Uncacheable)
-                    .unwrap();
-            }
-        }
+        let (inner, paddr_range) = if is_cache_coherent && !has_tdx {
+            let segment = FrameAllocOptions::new().alloc_segment(nframes)?;
+            let paddr_range = segment.paddr_range();
 
-        let start_daddr = match dma_type() {
-            DmaType::Direct => {
-                #[cfg(target_arch = "x86_64")]
-                crate::arch::if_tdx_enabled!({
-                    // SAFETY:
-                    //  - The address of a `USegment` is always page aligned.
-                    //  - A `USegment` always points to normal physical memory, so the address
-                    //    range falls in the GPA limit.
-                    //  - A `USegment` always points to normal physical memory, so all the pages
-                    //    are contained in the linear mapping.
-                    //  - The pages belong to a `USegment`, so they're all untyped memory.
-                    unsafe {
-                        tdx_guest::unprotect_gpa_range(paddr, frame_count).unwrap();
-                    }
-                });
-                paddr as Daddr
-            }
-            DmaType::Iommu => {
-                for i in 0..frame_count {
-                    let paddr = paddr + (i * PAGE_SIZE);
-                    // SAFETY: the `paddr` is restricted by the `paddr` and `frame_count` of the `segment`.
-                    unsafe {
-                        iommu::map(paddr as Daddr, paddr).unwrap();
-                    }
-                }
-                paddr as Daddr
-            }
+            (Inner::Segment(segment), paddr_range)
+        } else {
+            let (kva, paddr) = alloc_kva(nframes * PAGE_SIZE, has_tdx, is_cache_coherent)?;
```

Comment:
I think we can let `alloc_kva` accept `nframes`, then defer the overflow check to `FrameAllocOptions::alloc_segment_with`:
https://github.com/asterinas/asterinas/blob/049d9939cb893fcd7ecf95726ab6345fd2c31a99/ostd/src/mm/dma/util.rs#L67-L69

(I found that `FrameAllocOptions` does not currently check for overflows, but it's an irrelevant problem and more reasonable to fix in `FrameAllocOptions`.)